### DISCUSSION TITLE
WindTerm 2.5.0

### DISCUSSION THREAD

**Discussion by kingToolbox** [kingToolbox/WindTerm#921] [General]
# Spotlight

## Auto Completion
- [Intro Video: Auto Completion](https://kingtoolbox.github.io/2022/05/18/auto_completion/)
- [Intro Video: Auto Completion / History Commands](https://kingtoolbox.github.io/2022/05/18/auto_completion_history_commands/)
- [Intro Video: Auto Completion / Quick Commands](https://kingtoolbox.github.io/2022/05/18/auto_completion_quick_commands/)

#### Auto completion will recommend the most suitable parameters or phrases based on historical input, which greatly increases text input speed, because you don't have to memorize a large number of commands and type every letter in a command.

Auto completion supports:

- Linux Shell commands.
- MacOS Shell commands.
- Windows Cmd commands.
- PowerShell commands.
- Commands of any command line program, such as `Git`
- Quick Commands from the `QuickBar`

# Usage:
### Windows

|   Version | Architecture |  Operating Systems |
| --------- | -------------- | --------------------- |
| WindTerm_2.5.0_Windows_Portable_x86_32 | x86_32 | Windows 7, Windows 8, Windows 10, Windows 11 |
| WindTerm_2.5.0_Windows_Portable_x86_64 | x86_64 | Windows 7, Windows 8, Windows 10, Windows 11 |
| WindTerm_2.5.0_Mac_Portable_x86_64 | x86_64 | macOS 10.13, macOS 10.14, macOS 10.15, macOS 11, macOS 12 |
| WindTerm_2.5.0_Linux_Portable_x86_64 | x86_64 | Generic Linux (glibc 2.17 and later) |

#### Versions before Version 2.4.0

1. Start `WindTerm` and select the `profiles directory` and **quit**.
2. Replace the new `.wind/profiles` folder with the old `profiles` folder.
3. Restart `WindTerm` and all the sessions and configuration will be restored.

#### Versions after Version 2.4.0

1. Start `WindTerm` and select the `profiles directory` containing the old `.wind` folder.
2. It is recommended to choose the `Home Directory` to facilitate subsequent upgrades.

If the `.wind` folder needs to be moved to another system:
1. Quit `WindTerm`.
4. Replace the new `.wind` folder with the old `.wind` folder.
5. Restart `WindTerm` and all the sessions and configuration will be restored.

The **.wind** folder is cross-platform. 

# Known Issues:
- [ZMODEM] **Only support file names contains ASCII characters in Windows, otherwise the file name of the transfered file will be garbled after transmission, but it does not affect the correctness of the file content.** Linux and MacOS version does not have this problem.

# Translation:
- [CHINESE] Update the language file, contributed by **Lemonawa@github**. #246 #721
- [FRENCH] Update the language file, contributed by **EvoWebFrance@github**. #499 

# Features:
- [SESSION] **Supports auto completion for `Linux Shell`, `MacOS Shell`, `PowerShell`, `Windows Cmd`, `Git` and so on.** [Intro video](https://kingtoolbox.github.io/2022/05/18/auto_completion/)
  - **22** Git Commands
  - **32** Windows Cmd Commands
  - **42** Linux / MacOS Shell Commands
  - **3695** PowerShell Commands of `Windows Server 2022` and `PowerShell v7.3`
- [SESSION] **Supports auto completion for the command snippets from the `QuickBar`.** [Intro video](https://kingtoolbox.github.io/2022/06/20/auto_completion_quick_commands/)
- [SESSION] **Supports auto completion for `History Commands`.** [Intro video](https://kingtoolbox.github.io/2022/06/20/auto_completion_history_commands/)
- [SESSION] **Supports the encrypted storage of command history. The default maximum number of `history commands` is `10000`.**
- [SESSION] **Supports modifying the default configuration of `session`.** #775 #822 
- [SESSION] Supports ignoring bells received more than once within a specified second, the default is `2` seconds.
- [SCHEME] **Added `macOS` system, and all `Shell` sessions in the `macOS` system use the `macOS` by default.**
- [SSH] **Supports `Duplicate Session` and `Duplicate SSH Channel`.** #623 #669
- [SHELL] Automatically add `--login` and `-i` parameters when importing `Shell` sessions. #753 #781
- [SHELL] Automatically detect and set `LANG` environment variables. #347 #705
- [TERM] Supports customizing the key values sent by the `alt` modifier key and `meta` modifier key.
- [GUI] The filtering algorithm of the filter boxes is modified to a `multiple pattern matching algorithm`.
- [GUI] Added `Copy Command` menu item in the term view to quickly copy the command output.
- [GUI] Supports setting whether to enable `auto completion`, completion content, etc in `Settings Dialog - Terminal - Auto Completion`. #773 #784 #785 #786 #792 #811 #815 #860
- [GUI] Supports setting the number of visual rows of the `autocomplete list`, and whether to preselect the most matching item in `Settings Dialog - Terminal - Auto Completion`.
- [GUI] Supports setting whether to store `history command`, as well as storage size, storage time, etc. in `Settings Dialog - Terminal - Command History`.
- [GUI] Supports deleting the saved `history command` in `Settings Dialog - Terminal - Command History`. #773 
- [GUI] Supports setting the `bell style` in `Session Settings Dialog - Terminal - Bell`, including `Mute`, `Beep` and `Customized`, the default is `Beep`. #654 #655
- [GUI] Supports setting the `Locale` in `Session Settings Dialog - Shell - Process - Locale environment variables`. #347 #705
- [TAB] Restore the last modified tab name. #626 
- [TAB] Automatically add a number sequence to tabs with the same title for easy identification.
- [FILER] Add menu items `Copy directory path` and `Copy directory path to terminal`.

# Improvements:
- [SSH] **`Pageant` and `ssh-agent` authentication are no longer performed by default.**
- [SSH] **When the automatic login fails, the login will no longer restart, but will continue to log in manually.** #857 
- [SSH] When a disconnection is detected during the login, the `Reconnect` button will be displayed to facilitate a quick reconnection.
- [GUI] **Add the `Remember this step` check box to control whether to remember the authentication information of the current step when logging in.** #787 #857 
- [GUI] The filtering algorithm of the filter boxes prefers to match the first and consecutive characters.
- [GUI] Set the shortcut key for the action `Window.DuplicateView` to `Ctrl+Shift+D`. #770 
- [GUI] Set the shortcut key for the action `Terminal.OpenSession` to `Alt+O`.
- [GUI] Change the shortcut key for the action`New Session` from `Ctrl+N` to `Alt+N`. #782 #800 
- [GUI] Disabled the `Font Ligature` feature of most input boxes.
- [GUI] After clicking the `clear` authentication button, a message box will pop up to display the clear result.
- [GUI] In the `Session Settings` dialog, only the modified configuration items are written to the configuration file.
- [GUI] Replace all `Window.ShowPane*` with `Window.TogglePane*`. #880
- [SCHEME] `Cmd` supports highlighting `Slash Options`.
- [SCHEME] Impoved the highlighting of `Slash Options`.
- [SESSION] **The default lines of scrollback history is limited to `999999`.**
- [SESSION] New reconnection requests are ignored during the connection process. #765
- [SESSION] When using the mouse to select text in `vim`, the `alt` modifier key must be pressed simultaneously to enter free type mode. #897
- [SESSION] In the free type mode, pressing the `escape` key will exit the mode instead of sending the key to the remote host.
- [SESSION] Automatically executed commands are no longer trimmed. #870
- [SENDER] The scroll bar is no longer forced to scroll to the bottom of the screen when sending commands.

# Bugs:
- [SSH] **When creating a remote tunnel, the application will crash if the connected port is already in use.** #767
- [SSH] **Closing a remote tunnel will cause other tunnels in the same session to be closed by mistake.**
- [SSH] Cannot log in correctly when the password contains `!=`. #679 
- [GUI] The menu item `Session / Open` Session does not open the `Open Session` dialog correctly.
- [GUI] The `Listen Port` does not updated correctly when clicking on a different `tunnel` item in the `Tunnels` dialog.
- [GUI] In full screen mode, the drop-down box of the address bar will pop up in a new window instead of the current window. #804
- [GUI] When there is some wrapped lines, the terminal view may not be drawn correctly when the size changes.
- [LOG] Failed to create log files when the session name contains illegal characters, such as `:`, `|`, `<`, `>`, etc.
- [TAB] The text is not correctly elided according to the width of the `tab`.
- [TERM] Special characters cannot be entered using the `alt` modifier key in `macOS`.
- [SFTP] The configuration item `SSH - SFTP - Automatically fix invalid path characters` did not take effect when it was set to `false`.
- [SCHEME] Percentages are incorrectly identified as command prompts. #897
- [SESSION] `Clink` fails to start when the application path contains spaces. #736 #746
- [SESSION] Unable to enter the Cyrillic characters `р` and `Р`. #783
- [MODEM] The path pasted in the `sz` download dialog did not take effect. #836 

# Roadmap of v2.6 (Late-August, for reference only)
- **Resolve issues as much as possible**
- Improved Filer
- Improved Transfer
- Improved Quickbar
- SSH Agent Forwaring
- New memory allocator and manager and garbage collector. (Postponed to a later version )
- Command Snippet [Description](https://github.com/kingToolbox/WindTerm/issues/239#issuecomment-951934488)  (Postponed to a later version )
- SSH GSSAPI Authentication (Postponed to a later version )
- SSH Agent  (Postponed to a later version )
- Search in sessions. (Postponed to a later version )

# Intro Videos
- [Auto Completion](https://kingtoolbox.github.io/2022/05/18/auto_completion/)
- [Auto Completion / History Commands](https://kingtoolbox.github.io/2022/06/20/auto_completion_history_commands/)
- [Auto Completion / Quick Commands](https://kingtoolbox.github.io/2022/06/20/auto_completion_quick_commands/)

<hr /><em>This discussion was created from the release <a href='https://github.com/kingToolbox/WindTerm/releases/tag/2.5.0'>WindTerm 2.5.0</a>.</em>

**Comment by TheBlindM** [comment]
When executing some commands, the output will also be recorded as historical commands, such as df -h, I am curious how you get the historical commands

**Comment by wmashal** [comment]
I got this error when I am trying to open it in my mac 
“WindTerm” cannot be opened because the developer cannot be verified.

**Reply by bharadwaj-staked** [reply]
Looks like the macOS version isn't compatible with M2 chip.

**Comment by Azhun-HITwh** [comment]
How to use windterm on ubuntu 18.04? I got 
gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
when using tar -xzvf command to extract the tar.gz file

**Comment by jaukear** [comment]
WindTerm is the best SSH tool for me. I used many tools is not good!

**Comment by sipeiliu** [comment]
how to run it on linux?

**Comment by mikuUU-love** [comment]
哥们咱们就是能不能，优化一下命令历史记录偶尔显示偶尔不显示，而且会话不支持导出保存

**Comment by mikuUU-love** [comment]
哥们咱们就是能不能，优化一下命令历史记录偶尔显示偶尔不显示，而且会话不支持导出保存

**Reply by TheBlindM** [reply]
> 哥们咱们就是能不能，优化一下命令历史记录偶尔显示偶尔不显示，而且会话不支持导出保存

看一下T-Shell https://github.com/TheBlindM/T-Shell

**Comment by synohsu** [comment]
System:MacOS Sonoma 14.1.2 (23B92) 
OS language:Chinese
WindTerm:2.5.0
Session Windows:Serial
QS: can not display all，and display the “???”
![image](https://github.com/kingToolbox/WindTerm/assets/118065822/8930ed9e-1f83-41c7-87a7-5560cd6e84b2)

**Comment by andmanoppo** [comment]
The problem I encountered was that after filling in the server IP, I clicked to connect. It did not appear. Fill in the user name and password option. Please find the reason.
